pkg/mcp/internal/ratelimiter.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+// RateLimit is partially representing standard lib's rate limiter
+type RateLimit interface {
+	Wait(ctx context.Context) (err error)
+}
+
+// ConnectionRateLimit is an interface for per connection rate limiter
+type ConnectionRateLimit interface {
+	Wait(ctx context.Context, id string) (err error)
+	Remove(id string)
+}
+
+// RateLimiter is wrapper around golang's rate.Limit
+// it associates each limiter to a unique connection ID
+// for more balanced rate limiting among multiple connections
+type RateLimiter struct {
+	connectionFreq      time.Duration
+	connectionBurstSize int
+	perConnRateLimiter  map[string]RateLimit
+}
+
+// NewRateLimiter returns a new RateLimiter
+func NewRateLimiter(freq time.Duration, burstSize int) *RateLimiter {
```

Comment:
Anything under internal cannot be directly accessed. This may be okay for interfaces since users do not need to directly instantiate a `RateLimit` interface. However, something like `NewRateLimiter` won't be available.

Reply:
I put the rate limiting in internal intentionally since I couldn't envision a long term plan of it being used outside the mcp package, right now source/server should have no problem accessing it since they are under the same common root directory (nearby code).

Reply:
Ah, that makes sense. Given that it is internal, we can make it external later if necessary.